### PR TITLE
jobqueue: add the ability to dequeue by ID

### PR DESCRIPTION
**Pull request by ondrejbudai**
Needs to be done:

- [ ] tests
- [ ] db implementation
- [ ] documentation

This pull request includes:

- [ ] adequate testing for the new functionality or fixed issue
- [ ] adequate documentation informing people about the change such as
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

### CODE REVIEW

File: internal/jobqueue/fsjobqueue/fsjobqueue.go [hunk 1]
```diff
@@ -440,14 +481,14 @@ func selectUUIDChannel(ctx context.Context, chans []chan uuid.UUID) (uuid.UUID,
...
-			return uuid.Nil, ctx.Err()
+			return 0, ctx.Err()
 		} else {
-			return uuid.Nil, errors.New("channel was closed unexpectedly")
+			return 0, errors.New("channel was closed unexpectedly")
 		}
 	}
 
-	return value.Interface().(uuid.UUID), nil
+	return chosen - 1, nil
```
Comment (Gundersanne) [inline internal/jobqueue/fsjobqueue/fsjobqueue.go:561]: why -1?

Comment (Gundersanne) [reply internal/jobqueue/fsjobqueue/fsjobqueue.go:561]: ah nvm it counts from 1? weird.

**Comment by ondrejbudai** [comment]
This is now ready to be reviewed!

**Review by Gundersanne — APPROVED** [review]
Very nice :O ship it

**Review by teg — DISMISSED** [review]

File: internal/jobqueue/dbjobqueue/dbjobqueue.go [hunk 2]
```diff
@@ -42,6 +43,20 @@ const (
...
 
+	sqlDequeueByID = `
+		UPDATE jobs
+		SET token = $1, started_at = now()
+		WHERE id = (
+		  SELECT id
+		  FROM ready_jobs
+			  -- use ANY here, because "type in ()" doesn't work with bound parameters
+			  -- literal syntax for this is '{"a", "b"}': https://www.postgresql.org/docs/13/arrays.html
+		  WHERE id = ANY($2)
```
Comment (Gundersanne) [inline internal/jobqueue/dbjobqueue/dbjobqueue.go:54]: `WHERE id = $2`, ANY not needed

Also need to make sure started_at is zero; I tihnk the easiest way to do this is to compare started_at to the default constructed go time and pass that as $3.

File: internal/jobqueue/dbjobqueue/dbjobqueue.go [hunk 3]
```diff
@@ -208,6 +223,43 @@ func (q *dbJobQueue) Dequeue(ctx context.Context, jobTypes []string) (uuid.UUID,
...
+		return uuid.Nil, nil, "", nil, fmt.Errorf("error connecting to database: %v", err)
+	}
+
+	var jobType string
+	var args json.RawMessage
+	token := uuid.New()
+
+	err = conn.QueryRow(ctx, sqlDequeueByID, token, id).Scan(&token, &jobType, &args)
+	if err == pgx.ErrNoRows {
+		return uuid.Nil, nil, "", nil, fmt.Errorf("job '%v' isn't pending", id)
```
Comment (Gundersanne) [inline internal/jobqueue/dbjobqueue/dbjobqueue.go:243]: err needs to be `jobqueue.ErrNotPending`

**Review by Gundersanne — CHANGES_REQUESTED** [review]
base tests are failing

File: internal/jobqueue/dbjobqueue/dbjobqueue.go (hunk 2 above)
Comment (ondrejbudai) [reply internal/jobqueue/dbjobqueue/dbjobqueue.go]: Thanks! I fixed the where clause. `started_at` check is not needed, the `ready_jobs` view already has it.

File: internal/jobqueue/dbjobqueue/dbjobqueue.go (hunk 3 above)
Comment (ondrejbudai) [reply internal/jobqueue/dbjobqueue/dbjobqueue.go]: Right, fixed, thanks!

**Comment by ondrejbudai** [comment]
I made some mistakes in `dbjobqueue`, they should be now fixed.

**Review by Gundersanne — APPROVED** [review]

**Comment by ondrejbudai** [comment]
Yet another force-push. Dequeuing a job without deps returns `deps == nil` in `fsjobqueue` but an empty slice in case of `dbjobqueue`, I had to adjust tests because I expected `nil` in all cases.

**Review by teg — DISMISSED** [review]

**Review by Gundersanne — APPROVED** [review]

**Pull request merged**